--- a/net/optifine/gui/GuiMessage.java
+++ b/net/optifine/gui/GuiMessage.java
@@ -60,7 +60,7 @@
     {
         this.ticksUntilEnable = ticksUntilEnable;
 
-        for (Button button : this.getButtonList())
+        for (Button button : (List<Button>)(List<?>)this.getButtonList())
         {
             button.active = false;
         }
@@ -72,7 +72,7 @@
 
         if (--this.ticksUntilEnable == 0)
         {
-            for (Button button : this.getButtonList())
+            for (Button button : (List<Button>)(List<?>)this.getButtonList())
             {
                 button.active = true;
             }
