--- a/net/minecraft/client/gui/screens/ChatScreen.java
+++ b/net/minecraft/client/gui/screens/ChatScreen.java
@@ -219,7 +219,17 @@
         else
         {
             this.handleChatInput(this.input.getValue(), true);
-            this.minecraft.setScreen((Screen)null);
+            //Vivecraft
+            if (this.minecraft.vrSettings.seated)
+            {
+            	this.minecraft.setScreen((Screen)null);
+            }
+            else
+            {
+                this.input.setValue(this.initial);
+                this.historyPos = this.minecraft.gui.getChat().getRecentChat().size();
+            }
+            //
             return true;
         }
     }
