--- a/net/minecraft/client/gui/screens/SoundOptionsScreen.java
+++ b/net/minecraft/client/gui/screens/SoundOptionsScreen.java
@@ -12,6 +12,7 @@
 import net.minecraft.network.chat.Component;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.util.FormattedCharSequence;
+import org.vivecraft.settings.VRSettings;
 
 public class SoundOptionsScreen extends OptionsSubScreen
 {
@@ -39,7 +40,15 @@
                 ++k;
             }
         }
-
+        //Vivecraft
+        this.addRenderableWidget(new Button(this.width / 2 + i % 2 * 160, i + 22 * (k >> 1), 150, 20, Component.literal(this.minecraft.vrSettings.getButtonDisplayString(VRSettings.VrOptions.HRTF_SELECTION)), (p_213104_1_) ->
+        {
+        	this.clearWidgets();
+        	this.minecraft.vrSettings.setOptionValue(VRSettings.VrOptions.HRTF_SELECTION);
+        	this.minecraft.vrSettings.saveOptions();
+        	this.init();
+        }));
+        //
         if (k % 2 == 1)
         {
             ++k;
