--- a/net/minecraft/world/item/crafting/RecipeManager.java
+++ b/net/minecraft/world/item/crafting/RecipeManager.java
@@ -31,6 +31,8 @@
 import net.minecraft.world.level.Level;
 import org.slf4j.Logger;
 
+import org.vivecraft.utils.ASMInjections;
+
 public class RecipeManager extends SimpleJsonResourceReloadListener
 {
     private static final Gson GSON = (new GsonBuilder()).setPrettyPrinting().disableHtmlEscaping().create();
@@ -68,7 +70,10 @@
                 LOGGER.error("Parsing error loading recipe {}", resourcelocation, jsonparseexception);
             }
         }
-
+        
+        //Vivecraft vanilla only - forge use ASM
+        ASMInjections.injectItems(map, builder);
+        
         this.recipes = map.entrySet().stream().collect(ImmutableMap.toImmutableMap(Map.Entry::getKey, (p_44033_) ->
         {
             return p_44033_.getValue().build();
