--- a/net/minecraft/client/Minecraft.java
+++ b/net/minecraft/client/Minecraft.java
@@ -240,11 +240,13 @@
 import org.apache.commons.io.FileUtils;
 import org.lwjgl.util.tinyfd.TinyFileDialogs;
 import org.slf4j.Logger;
+import org.vivecraft.render.PlayerModelController;
 
 public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements WindowEventHandler
 {
     static Minecraft instance;
     private static final Logger LOGGER = LogUtils.getLogger();
+    public final String minecriftVerString = "Vivecraft 1.19.2  jrbudda-NONVR-2-b2";
     public static final boolean ON_OSX = Util.getPlatform() == Util.OS.OSX;
     private static final int MAX_TICKS_PER_UPDATE = 10;
     public static final ResourceLocation DEFAULT_FONT = new ResourceLocation("default");
@@ -633,7 +635,7 @@
 
     private String createTitle()
     {
-        StringBuilder stringbuilder = new StringBuilder("Minecraft");
+        StringBuilder stringbuilder = new StringBuilder(minecriftVerString);
 
         if (checkModStatus().shouldReportAsModified())
         {
@@ -2130,6 +2132,8 @@
             this.gameRenderer.shutdownEffect();
         }
 
+        PlayerModelController.getInstance().tick();
+        
         if (!this.pause)
         {
             this.musicManager.tick();
