--- a/net/optifine/override/PlayerControllerOF.java
+++ b/net/optifine/override/PlayerControllerOF.java
@@ -12,6 +12,7 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.EntityHitResult;
+import org.vivecraft.api.NetworkHelper;
 
 public class PlayerControllerOF extends MultiPlayerGameMode
 {
@@ -45,13 +46,24 @@
     public InteractionResult useItem(Player player, InteractionHand hand)
     {
         this.acting = true;
+        //Vivecraft
+        NetworkHelper.overrideLook(player, Minecraft.getInstance().vrPlayer.getRightClickLookOverride(player, hand.ordinal()));
+        //
         InteractionResult interactionresult = super.useItem(player, hand);
         this.acting = false;
         return interactionresult;
     }
+    //Vivecraft
+    public void releaseUsingItem(Player pPlayer)
+    {
+        NetworkHelper.overrideLook(pPlayer, Minecraft.getInstance().vrPlayer.getRightClickLookOverride(pPlayer, pPlayer.getUsedItemHand().ordinal()));
+        super.releaseUsingItem(pPlayer);
+    }
+    //
 
     public InteractionResult useItemOn(LocalPlayer player, InteractionHand hand, BlockHitResult rayTrace)
     {
+        NetworkHelper.overrideLook(player, rayTrace.getLocation().subtract(player.getEyePosition(1.0F)).normalize());
         this.acting = true;
         this.lastClickBlockPos = rayTrace.getBlockPos();
         InteractionResult interactionresult = super.useItemOn(player, hand, rayTrace);
