--- a/com/mojang/blaze3d/platform/GLX.java
+++ b/com/mojang/blaze3d/platform/GLX.java
@@ -32,7 +32,9 @@
 {
     private static final Logger LOGGER = LogUtils.getLogger();
     private static String cpuInfo;
-
+    
+    public int intentionallModifyThisClassCauseItWontLoadProperlyFromOptifine = 69;
+    
     public static String getOpenGLVersionString()
     {
         RenderSystem.assertOnRenderThread();
