--- a/net/minecraft/server/packs/VanillaPackResources.java
+++ b/net/minecraft/server/packs/VanillaPackResources.java
@@ -247,7 +247,7 @@
             stream.filter((p_215350_0_) ->
             {
                 return !p_215350_0_.endsWith(".mcmeta") && Files.isRegularFile(p_215350_0_);
-            }).mapMulti((p_242539_2_, p_242539_3_) ->
+            }).<ResourceLocation>mapMulti((p_242539_2_, p_242539_3_) ->
             {
                 String s = path.relativize(p_242539_2_).toString().replaceAll("\\\\", "/");
                 ResourceLocation resourcelocation = ResourceLocation.tryBuild(pType, s);
