--- a/net/minecraft/client/gui/screens/TitleScreen.java
+++ b/net/minecraft/client/gui/screens/TitleScreen.java
@@ -544,10 +544,6 @@
 
     static record WarningLabel(Font font, MultiLineLabel label, int x, int y)
     {
-        private final Font font;
-        private final MultiLineLabel label;
-        private final int x;
-        private final int y;
         public void render(PoseStack p_232791_, int p_232792_)
         {
             this.label.renderBackgroundCentered(p_232791_, this.x, this.y, 9, 2, 1428160512);
